File: backend/GameApp.Domain/Entities/Game.cs
using GameApp.Domain.Entities.Scenes;
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Characters;

namespace GameApp.Domain.Entities;

public class Game
{
    private readonly Guid Id;
    private readonly GameDifficulty Difficulty;
    private readonly Character Character;
    private readonly int NumberScenesToFinish;
    private readonly List<Scene> ListCompletedScenes;
    private readonly List<Scene> ListCurrentScenes;
    private readonly List<UserAction> ListCurrentUserActions;  // all possible actions that the user can perform in the current scenes
    private readonly FinalScene FinalScene;
    private readonly GameStatus Status;
    private readonly Enemy? CurrentEnemy;  // could be an enemy on current scene

    ///-------------CREATE ATRIBUTE: CurrentEnemy

    // // Required for EF Core
    // private Game()
    // {
    //     Id = Guid.NewGuid();
    //     Character = null!;
    //     ListCompletedScenes = new List<Scene>();
    //     ListCurrentScenes = new List<Scene>();
    //     ListCurrentUserActions = new List<UserAction>();
    //     FinalScene = null!;
    //     Status = GameStatus.GameInProgress;
    // }

    // Default constructor
    public Game(GameDifficulty difficulty, Character character, int numberScenesToFinish, FinalScene finalScene, List<Scene> listCurrentScenes, List<UserAction> listCurrentUserActions)
    {
        Id = Guid.NewGuid();
        Difficulty = difficulty;
        Character = character;
        NumberScenesToFinish = numberScenesToFinish;
        FinalScene = finalScene;
        ListCompletedScenes = new List<Scene>();
        ListCurrentScenes = listCurrentScenes;
        ListCurrentUserActions = listCurrentUserActions;
        Status = GameStatus.GameInProgress;
    }

    // Restore constructor
    public Game(Guid id, GameDifficulty difficulty, Character character, int numberScenesToFinish, List<Scene> completedScenes, FinalScene finalScene, List<Scene> listCurrentScenes, List<UserAction> listCurrentUserActions, GameStatus status, Enemy? currentEnemy)
    {
        Id = id;
        Difficulty = difficulty;
        Character = character;
        NumberScenesToFinish = numberScenesToFinish;
        ListCompletedScenes = new List<Scene>(completedScenes);
        FinalScene = finalScene;
        ListCurrentScenes = listCurrentScenes;
        ListCurrentUserActions = listCurrentUserActions;
        CurrentEnemy = currentEnemy;
        Status = status;
    }

    // Getters
    public Guid GetGuid() => Id;

    public GameDifficulty GetDifficulty() => Difficulty;
    public Character GetCharacter() => Character;
    public int GetNumberScenesToFinish() => NumberScenesToFinish;
    public List<Scene> GetCompletedScenes() => new List<Scene>(ListCompletedScenes);
    public FinalScene GetFinalScene() => FinalScene;
    public List<Scene> GetCurrentScenes() => new List<Scene>(ListCurrentScenes);
    public List<UserAction> GetCurrentUserAction() => new List<UserAction>(ListCurrentUserActions);
    public GameStatus GetGameStatus() => Status;
    public Enemy? GetCurrentEnemy() => CurrentEnemy;

    // Setters
    public Game SetCharacter(Character newCharacter) =>
        new Game(Id, Difficulty, newCharacter, NumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);

    public Game SetNumberScenesToFinish(int newNumberScenesToFinish) =>
        new Game(Id, Difficulty, Character, newNumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);

    public Game AddCompletedScene(Scene newScene)
    {
        var newList = new List<Scene>(ListCompletedScenes) { newScene };
        return new Game(Id, Difficulty, Character, NumberScenesToFinish, newList, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);
    }

    public Game RemoveLastCompletedScene()
    {
        if (ListCompletedScenes.Count == 0) return this;
        var newList = new List<Scene>(ListCompletedScenes);
        newList.RemoveAt(newList.Count - 1);
        return new Game(Id, Difficulty, Character, NumberScenesToFinish, newList, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);
    }

    public Game SetFinalScene(FinalScene newFinalScene) =>
        new Game(Id, Difficulty, Character, NumberScenesToFinish, ListCompletedScenes, newFinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);

    public Game SetCurrentScenes(List<Scene> newListCurrentScenes) =>
        new Game(Id, Difficulty, Character, NumberScenesToFinish, ListCompletedScenes, FinalScene, newListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);
        
    public Game SetCurrentUserActions(List<UserAction> newListCurrentUserActions) =>
        new Game(Id, Difficulty, Character, NumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, newListCurrentUserActions, Status, CurrentEnemy);

    public Game SetCurrentEnemy(Enemy? newCurrentEnemy) =>
        new Game(Id, Difficulty, Character, NumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, newCurrentEnemy);
    
    public Game SetGameStatus(GameStatus newStatus) =>
        new Game(Id, Difficulty, Character, NumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, ListCurrentUserActions, newStatus, CurrentEnemy);

    public Game SetDifficulty(GameDifficulty newDifficulty) =>
        new Game(Id, newDifficulty, Character, NumberScenesToFinish, ListCompletedScenes, FinalScene, ListCurrentScenes, ListCurrentUserActions, Status, CurrentEnemy);


    public Game UpdateGame(GameDifficulty newDifficulty, Character newCharacter, int newNumberScenesToFinish, List<Scene> newCompletedScenes, FinalScene newFinalScene, List<Scene> newListCurrentScenes, List<UserAction> newlistCurrentUserActions, GameStatus newStatus, Enemy? newCurrentEnemy) =>
    new Game(Id, newDifficulty, newCharacter, newNumberScenesToFinish, newCompletedScenes, newFinalScene, newListCurrentScenes, newlistCurrentUserActions, newStatus, newCurrentEnemy);

    // To string
    public override string ToString()
    {
        string completedScenesStr = string.Join(", ", ListCompletedScenes.ConvertAll(s => s.GetName().ToString()));
        string currentScenesStr = string.Join(", ", ListCurrentScenes.ConvertAll(s => s.GetName().ToString()));
        string currentUserActionsStr = string.Join(", ", ListCurrentUserActions.ConvertAll(s => s.ToString()));
        return $"Game {Id}(Difficulty:{Difficulty}): Character={Character.GetName()}, NumberScenesToFinish={NumberScenesToFinish}, " +
               $"CompletedScenes=[{completedScenesStr}], FinalScene={FinalScene?.GetName()}, CurrentScenes=[{currentScenesStr}], " + 
               $"CurrentUserAction=[{currentUserActionsStr}], GameStatus={Status}, CurrentEnemy={CurrentEnemy}";
    }
}

File: frontend/gameapp-frontend/src/domain/entities/game.js
import { UserAction } from '../enumerates/user-action.js';
import { Character } from '../value-objects/characters/character.js';
import { Enemy } from './enemy.js';
import { NothingHappensScene } from './scenes/nothing-happens-scene.js';
import { Scene } from './scenes/scene.js';
import { validateOrGenerateUUID } from '../../utils/validate-or-generate-uuid.js';

export class Game {
    constructor(character, numberScenesToFinish, finalScene, listCurrentScenes = [], listCurrentUserActions = [], listCompletedScenes = [], currentEnemy = null, id = null) {
        this._id = validateOrGenerateUUID(id);

        if (!(character instanceof Character)) throw new TypeError("character must be an instance of Character");
        this._character = character;

        if (typeof numberScenesToFinish !== "number" || !Number.isInteger(numberScenesToFinish) || numberScenesToFinish <= 0) {
            throw new TypeError("numberScenesToFinish must be an integer greater than 0");
        }
        this._numberScenesToFinish = numberScenesToFinish;


        if (!(finalScene instanceof NothingHappensScene)) throw new TypeError("finalScene must be an instance of NothingHappensScene");
        this._finalScene = finalScene;

        this.#validateSceneList(listCurrentScenes, "listCurrentScenes");
        this._listCurrentScenes = [...listCurrentScenes];

        this.#validateSceneList(listCompletedScenes, "listCompletedScenes");
        this._listCompletedScenes = [...listCompletedScenes];

        this.#validateUserActionsList(listCurrentUserActions, "listCurrentUserActions");
        this._listCurrentUserActions = listCurrentUserActions;

        if (currentEnemy != null && !(currentEnemy instanceof Enemy)) throw new TypeError("currentEnemy must be an instance of Enemy or null");
        this._currentEnemy = currentEnemy;
    }

    #validateSceneList(list, paramName) {
        if (!Array.isArray(list)) throw new TypeError(`${paramName} must be an array`);
        for (const scene of list) {
            if (!(scene instanceof Scene)) {
                throw new TypeError(`All elements of ${paramName} must be instances of Scene or its subclasses`);
            }
        }
    }

    #validateUserActionsList(list, paramName) {
        if (!Array.isArray(list)) {
            throw new TypeError(`${paramName} must be an array`);
        }

        const validActions = Object.values(UserAction);

        for (const action of list) {
            if (!validActions.includes(action)) {
                throw new TypeError(
                    `${paramName} contains an invalid UserAction value: '${action}'`
                );
            }
        }
    }

    // getter
    get id() { return this._id; }
    get character() { return this._character; }
    get numberScenesToFinish() { return this._numberScenesToFinish; }
    get completedScenes() { return [...this._listCompletedScenes]; }
    get finalScene() { return this._finalScene; }
    get currentScenes() { return [...this._listCurrentScenes]; }
    get currentUserActions() { return [...this._listCurrentUserActions]; }
    get currentEnemy() { return this._currentEnemy; }

    // setter
    setCharacter(newCharacter) {
        return new Game(newCharacter, this._numberScenesToFinish, this._finalScene, this._listCurrentScenes, this._listCurrentUserActions, this._listCompletedScenes, this._currentEnemy, this._id);
    }

    setNumberScenesToFinish(newNumber) {
        return new Game(this._character, newNumber, this._finalScene, this._listCurrentScenes, this._listCurrentUserActions, this._listCompletedScenes, this._currentEnemy, this._id);
    }

    addCompletedScene(newScene) {
        const newList = [...this._listCompletedScenes, newScene];
        return new Game(this._character, this._numberScenesToFinish, this._finalScene, this._listCurrentScenes, this._listCurrentUserActions, newList, this._currentEnemy, this._id);
    }

    removeLastCompletedScene() {
        if (this._listCompletedScenes.length === 0) return this;
        const newList = [...this._listCompletedScenes];
        newList.pop();
        return new Game(this._character, this._numberScenesToFinish, this._finalScene, this._listCurrentScenes, this._listCurrentUserActions, newList, this._currentEnemy, this._id);
    }

    setFinalScene(newFinalScene) {
        return new Game(this._character, this._numberScenesToFinish, newFinalScene, this._listCurrentScenes, this._listCurrentUserActions, this._listCompletedScenes, this._currentEnemy, this._id);
    }

    setCurrentScenes(newList) {
        return new Game(this._character, this._numberScenesToFinish, this._finalScene, newList, this._listCurrentUserActions, this._listCompletedScenes, this._currentEnemy, this._id);
    }

    setCurrentUserActions(newListCurrenUserActions) {
        return new Game(this._character, this._numberScenesToFinish, this._finalScene, this._listCurrentScenes, newListCurrenUserActions, this._listCompletedScenes, this._currentEnemy, this._id);
    }

    setCurrentEnemy(newCurrentEnemy) {
        return new Game(this._character, this._numberScenesToFinish, this._finalScene, this._listCurrentScenes, this._listCurrentUserActions, this._listCompletedScenes, newCurrentEnemy, this._id);
    }

    updateGame(newCharacter, newNumberScenesToFinish, newCompletedScenes, newFinalScene, newListCurrenUserActions, newListCurrentScenes, newCurrentEnemy) {
        return new Game(newCharacter, newNumberScenesToFinish, newFinalScene, newListCurrentScenes, newListCurrenUserActions, newCompletedScenes, newCurrentEnemy, this._id);
    }

    toString() {
        const completedStr = this._listCompletedScenes.map(s => s.name.toString()).join(", ");
        const currentScenesStr = this._listCurrentScenes.map(s => s.name.toString()).join(", ");
        const currentUserActionsStr = this._listCurrentUserActions.join(", ");
        return `Game ${this._id}: Character=${this._character.name.toString()}, NumberScenesToFinish=${this._numberScenesToFinish}, ` +
            `CompletedScenes=[${completedStr}], FinalScene=${this._finalScene?.name.toString()}, CurrentScenes=[${currentScenesStr}], CurrentUserActions=[${currentUserActionsStr}], CurrentEnemy=${this._currentEnemy}`;
    }
}
